<div class="alert alert-info">
Привет, Анастасия! Меня зовут Светлана и я буду проверять твой проект. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе. Предлагаю общаться на «ты». Но если это не удобно - дай знать, и мы перейдем на «вы».

<div class="alert alert-success">
<b>👍 Успех:</b> Зелёным цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.
</div>
<div class="alert alert-warning">
<b>🤔 Рекомендация:</b> Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).
</div>
<div class="alert alert-danger">
<b>😔 Необходимо исправить:</b> Красным цветом выделены комментарии, без исправления которых, я не смогу принять проект :(
</div>
<div class="alert alert-info">
<b>👂 Совет:</b> Какие-то дополнительные материалы
</div>
Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом.
Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Например, вот так.</div>
Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта быстрее.
 </div>

# ПРОЕКТ SQL: ЧТЕНИЕ ПО ПОДПИСКЕ

Цель проекта: сформулировать ценностное предложение для нового продукта для сервиса чтения книг по подписке.

Для этого нам предоставлена база данных с информацией о книгах, издательствах, авторах, пользовательских обзоров книг, уже имеющихся в купленном сервисе.

Для начала импортируем необходимые библиотеки и напишем путь к базе данных

In [1]:
#lets import necessary libraries
import pandas as pd
from sqlalchemy import text, create_engine
from PIL import Image

#set parameters
db_config = {'user': 'praktikum_student', #user name
'pwd': 'Sdf4$2;d-d30pp', # password
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', #host
'port': 6432, #connection port
'db': 'data-analyst-final-project-db'} #database name
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
#save connector
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

 Посмотрим первые 5 строк всех таблиц из базы данных:
 **Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [2]:
query = '''SELECT *
             FROM books
             limit 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [3]:
query = '''SELECT *
             FROM authors
             limit 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

In [4]:
query = '''SELECT *
             FROM publishers
             limit 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [5]:
query = '''SELECT *
             FROM ratings
             limit 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [6]:
query = '''SELECT *
             FROM reviews
             limit 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Отлично, все таблицы просмотрены, теперь осталось вывести схему базы данных, чтобы под рукой были все взаимосвязи и ключи между таблицами

<a href="https://ibb.co/kXTfmmp">
<img src="https://i.ibb.co/w4qXccT/Untitled-33.png" alt="Untitled-33" border="0">
</a>

 <div class="alert alert-success">
 <b>👍 Успех:</b> Есть описание задачи, схема БД, импортированы все нужные библиотеки, создано подключение к БД, просмотрены таблицы
 </div>

### Задания для анализа

1) Посчитайте, сколько книг вышло после 1 января 2000 года;

In [7]:
#how many books issued after 1/1/2000
query = '''SELECT COUNT(book_id)
             FROM books
             WHERE CAST (publication_date AS date) > '2000-01-01'
             '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

**Вывод**: после 1 января 2000 года вышло 819 книг

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

2) Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [8]:
#count number of reviews and average score
query = '''SELECT b.book_id AS id, b.title AS title, COUNT (DISTINCT rw.review_id) as number_of_reviewes, \
                                                                    ROUND (AVG(rt.rating),2) as average_rating
             FROM books AS b 
             FULL OUTER JOIN reviews AS rw ON b.book_id=rw.book_id
             JOIN ratings AS rt ON b.book_id=rt.book_id
             GROUP BY 1,2
             ORDER BY number_of_reviewes DESC, average_rating DESC
             '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

id                                              title  \
0    948                            Twilight (Twilight  #1)   
1    302  Harry Potter and the Prisoner of Azkaban (Harr...   
2    299  Harry Potter and the Chamber of Secrets (Harry...   
3    656                                     The Book Thief   
4    734                                   The Glass Castle   
..   ...                                                ...   
995  191  Disney's Beauty and the Beast (A Little Golden...   
996  387                               Leonardo's Notebooks   
997  221                          Essential Tales and Poems   
998   83    Anne Rice's The Vampire Lestat: A Graphic Novel   
999  808                            The Natural Way to Draw   

     number_of_reviewes  average_rating  
0                     7            3.66  
1                     6            4.41  
2                     6            4.29  
3                     6            4.26  
4                     6            4.21  
..                  ...             ...  
995                   0            4.00  
996                   0            4.00  
997                   0            4.00  
998                   0            3.67  
999                   0            3.00  

[1000 rows x 4 columns]

**Вывод**: получена таблица со всеми книгами, количеством отзывов и средним рейтингом по каждой книге. Наибольшее количество обзоров у книги "Сумерки"

 <div class="alert alert-danger">
 <s><b>😔 Необходимо исправить:</b> Ты группируешь по неуникальному признаку, у нас всего 1000 книг</s>
 </div>
 
 <div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Исправлено.</div>
 
  <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

3) Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;


In [9]:
#count publisher with most quantity of books over 50 pages
query = '''SELECT p.publisher, COUNT (DISTINCT b.book_id) as number_of_books
             FROM publishers AS p
             JOIN books AS b ON p.publisher_id=b.publisher_id
             WHERE num_pages > 50
             GROUP BY p.publisher
             ORDER BY number_of_books DESC
             LIMIT 1
             '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  number_of_books
0  Penguin Books               42

**Вывод**: издательство, выпустившее наибольшее число книг толще 50 страниц - Penguin Books. Оно выпустило 42 толстых книги.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

4) Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [12]:
query = '''SELECT a.author, ROUND(AVG(r.rating),2)
FROM authors AS a
LEFT JOIN books AS b ON a.author_id=b.author_id
LEFT JOIN ratings AS r on b.book_id=r.book_id
WHERE b.book_id IN (SELECT b.book_id
                    FROM books AS b
                    LEFT JOIN ratings AS r on b.book_id=r.book_id
                    GROUP BY b.book_id
                    HAVING COUNT (r.rating_id) >= 50)
GROUP BY 1
ORDER BY AVG(r.rating) desc
limit 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  round
0  J.K. Rowling/Mary GrandPré   4.29

**Вывод**: писательница Джоанн Роулинг и иллюстратор Мэри ГрандПре, с их детищем Гарри Поттером, лидируют по среднему рейтингу среди наиболее часто оцениваемых книг.  

 <div class="alert alert-danger">
 <s><b>😔 Необходимо исправить:</b> В подзапросе ты оставляешь только одну книгу, а нам нужно найти среднюю оценку по всем книгам авторов и выбрать автора с самой высокой средней оценкой по всем книгам, попадающим под условие. То есть тебе нужно еще раз сгруппировать данные по автору и найти среднее от среднего</s>
 </div>
 
  <div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Исправлено.</div>
  
   <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

5) Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [11]:
#count average number of reviewes from users that made more 48 ratings
query = '''WITH
rev AS (SELECT username, COUNT (DISTINCT review_id) AS number_of_reviews
            FROM reviews
            GROUP BY username),
rat AS (SELECT username, COUNT (rating) 
            FROM ratings
            GROUP BY username
            HAVING COUNT (rating) > 48)
SELECT ROUND(AVG(rev.number_of_reviews))
FROM rev JOIN rat ON rev.username = rat.username
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

round
0   24.0

**Вывод**: активные пользователи, которые поставили больше 48 оценок, довольно активны и в написании обзоров - в среднем они сделали 24 обзора по книгам приложения

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

**ОБЩИЙ ВЫВОД**: 
Полученные данные прочитаны и проведен анализ для формирования ценностного предложения для нового продукта:
- согласно нашим базам данных, после 1 января 2000 года вышло в свет 819 новых книг. Исходя из этих данных можно строить прогнозы по продуктивности авторов на будущие годы, и соответственно планировать свои доходы
- собраны данные по каждой книге по количеству обзоров и средней оценке. Пользуясь этими данными, можно вычислить какие книги наиболее интересны пользователям - по таким книгам пишут больше обзоров и ставят максимальные рейтинги
- определено издательство, выпустившее наибольшее число книг толще 50 страниц - это Penguin Books. Вот и стратегический партнер для нашего нового бизнеса! Можно продумать кросс-маркетинговые акции либо программы сотрудничества с этим издательством.
- выявлен автор с самой высокой средней оценкой книг среди книг с 50 и более оценками: это всемирно известная писательница Джоанн Роулинг и иллюстратор ее книг о Гарри Поттере - Мэри ГрандПре. Опять же, данная информация помогает лучше узнать целевую аудиторию - что нравится читателям нашего сервиса.
- среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24. Это говорит о том, что активные пользователи в среднем пишут в два раза меньше отзывов, чем ставят оценок.

 <div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Ошибки исправлены, вывод не поменялся.</div>

 <div class="alert alert-success">
 <b>👍 Успех:</b> Хорошая работа, поздравляю с окончанием курса! У тебя получились уверенные работы на финальном спринте, логичные и последовательные, ты умешшь рассуждать, визуализировать данные и делать выводы, это очень хорошие качества для аналитика! Рекомендую еще пройти бесплатный курс от практикума по <a href='https://practicum.yandex.ru/profile/ycloud-datalens/'> DataLens<a>
 </div>